src/client/dfuse/pil4dfs/int_dfs.c

```diff
@@ -3310,70 +3335,284 @@ new_readdir(DIR *dirp)
- *new_execve(const char *filename, char *const argv[], char *const envp[])
- *{
- *	printf("DBG> execve(%s)\n", filename);
- *	return real_execve(filename, argv, pre_envp(envp));
- *}
- *
- *static int
- *new_execvp(const char *filename, char *const argv[])
- *{
- *	printf("DBG> execvp(%s)\n", filename);
- *	return real_execve(filename, argv, pre_envp(NULL));
- *}
- *
- *static int
- *new_execv(const char *filename, char *const argv[])
- *{
- *	printf("DBG> execv(%s)\n", filename);
- *	pre_envp(NULL);
- *	return real_execve(filename, argv, pre_envp(NULL));
- *}
- *
- *static pid_t
- *new_fork(void)
- *{
- *	pid_t pid;
- *	pid = real_fork();
- *	if (pid) {
- *		// parent process: do nothing
- *		return pid;
- *	} else {
- *		init_dfs();
- *		return pid;
- *	}
- *}
+extern char **__environ;
+
+/* This is the number of environmental variables that would be forced to set in child process.
+ * "LD_PRELOAD" is a special case and it is not included in the list.
  */
+static char  *env_list[] = {"D_IL_REPORT", "DAOS_MOUNT_POINT", "DAOS_POOL", "DAOS_CONTAINER",
+			    "D_IL_MAX_EQ", "D_LOG_FILE", "D_IL_ENFORCE_EXEC_ENV",  "DD_MASK",
+			    "DD_SUBSYS", "D_LOG_MASK"};
+
+/* Environmental variables could be cleared in some applications. To make sure all libpil4dfs
+ * related env properly set, we intercept execve and its variants to check envp[] and append our
+ * env when needed.
+ * pre_envp() scans provided envp to look for libpil4dfs related envs (env_list[]) and returns a
+ * new envp with missing libpil4dfs related envs appended.
+ * The fd for daos logging may be not valid here, so printf() is used in this function.
+ */
+static char **
+pre_envp(char *const envp[])
+{
+	int    i, j, rc;
+	int    len, len2, len_total;
+	int    num_env_append;
+	char **new_envp;
+	char  *pil4df_path;
+	char  *new_preload_str;
+	int    num_entry       = 0;
+	int    num_entry_found = 0;
+	int    idx_preload     = -1;
+	bool   preload_included                   = false;
+	bool   pil4dfs_in_preload                 = false;
+	bool   env_found[ARRAY_SIZE(env_list)]    = {false};
+	/* the buffer allocated for the env string to append */
+	char  *env_buf_list[ARRAY_SIZE(env_list)] = {NULL};
+	char  *env_value                          = NULL;
+
+	/* Including env_list[] and "LD_PRELOAD" */
+	num_env_append = ARRAY_SIZE(env_list) + 1;
+
+	if (envp == NULL) {
+		num_entry = 0;
+	} else if (envp[0] == NULL) {
+		num_entry = 0;
+	} else {
+		while (envp[num_entry]) {
+			/* scan the env in env_list[] to check whether they exist or not */
+			if (strncmp(envp[num_entry], STR_AND_SIZE_M1("LD_PRELOAD")) == 0) {
+				preload_included = true;
+				idx_preload      = num_entry;
+				num_entry_found++;
+				if (strstr(envp[num_entry], "libpil4dfs.so"))
+					pil4dfs_in_preload = true;
+			}
+			/* The list of env is not too long. We use a simple loop to lookup for
+			 * simplicity. This function is not performance critical.
+			 */
+			for (i = 0; i < ARRAY_SIZE(env_list); i++) {
```

Comment:
There's a case here where this code will be invoked but not required to.  If LD_PRELOAD is being used but one of the env vars you're testing for is not set then this will be detected and the code will start to modify the env, but if the env isn't set then it won't be able to be copied so in effect this will make a copy but then not modify it.

To catch this case then I think this while code should loop and call d_agetenv_str() first for each entry and then only search the list for ones which are set in the parent.

Reply:
Thank you very much! Revised as you suggested.